### ISSUE TITLE
Why was the adversarial example obtained by MIPVerify not considered a genuine counterexample when incorporated into the actual reasoning process?

### ISSUE THREAD

**Issue by ace-shifu** [vtjeng/MIPVerify.jl#158]
I utilized MIPVerify to generate an adversarial example for the MNIST dataset. Here are the findings: I acquired an adversarial example through MIPVerify. However, during the actual inference process, I observed that the resulting label persisted as 8.

![screen1](https://github.com/vtjeng/MIPVerify.jl/assets/62993420/0fbb5302-02f1-4c3f-839e-1e48a39f754c)

**Comment by vtjeng** [comment]
Hi there, I think this might be an off-by-one issue between the labels and
the indices. Could you try call `find_adversarial_example` with a target
list of `[1]`, and check what index you get? I expect the index would be 0.

**Comment by ace-shifu** [comment]
Sorry, but it still didn't work.  It appears that the perturbations had no effect on the network.
![target1](https://github.com/vtjeng/MIPVerify.jl/assets/62993420/1d8fa975-2b19-4a36-8222-124a78049c4d)

**Comment by ace-shifu** [comment]
> Sorry, but it still didn't work. It appears that the perturbations had no effect on the network. ![target1](https://private-user-images.githubusercontent.com/62993420/312327154-1d8fa975-2b19-4a36-8222-124a78049c4d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzMTA0NDUsIm5iZiI6MTcxMDMxMDE0NSwicGF0aCI6Ii82Mjk5MzQyMC8zMTIzMjcxNTQtMWQ4ZmE5NzUtMmIxOS00YTM2LTgyMjItMTI0YTc4MDQ5YzRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEzVDA2MDkwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2ZjMyMDEwMmU2Zjk2MDg5ZDQxYWU0MDVjMTY1NWQyZjkyYWY0MWQwYmJmZGQxN2YxMDcwYjI0ZmI4MWNmOTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zqkd5_L-nu_6XRtL5oQJXk-LNbAmtuoLLy58ooEIh4E)

I comprehend that in Julia, arrays commence with an index of 1, rather than zero. Upon examination, I ascertain that the genuine label assigned to this image is 7, with its corresponding index in julia code being 8.
![Screenshot from 2024-03-13 14-11-04](https://github.com/vtjeng/MIPVerify.jl/assets/62993420/19d27e1f-74aa-43cd-ae84-fc2d802c63f3)

**Comment by vtjeng** [comment]
> > Sorry, but it still didn't work. It appears that the perturbations had no effect on the network. ![target1](https://private-user-images.githubusercontent.com/62993420/312327154-1d8fa975-2b19-4a36-8222-124a78049c4d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzMTA0NDUsIm5iZiI6MTcxMDMxMDE0NSwicGF0aCI6Ii82Mjk5MzQyMC8zMTIzMjcxNTQtMWQ4ZmE5NzUtMmIxOS00YTM2LTgyMjItMTI0YTc4MDQ5YzRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEzVDA2MDkwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2ZjMyMDEwMmU2Zjk2MDg5ZDQxYWU0MDVjMTY1NWQyZjkyYWY0MWQwYmJmZGQxN2YxMDcwYjI0ZmI4MWNmOTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zqkd5_L-nu_6XRtL5oQJXk-LNbAmtuoLLy58ooEIh4E)
> 
> I comprehend that in Julia, arrays commence with an index of 1, rather than zero. Upon examination, I ascertain that the genuine label assigned to this image is 7, with its corresponding index in julia code being 8.

Thanks! Unfortunately I no longer have an academic license for Gurobi, so verifying this part of the code is going to take me some time. In the meantime, are you able to share the output of the following?

```
# inspecting the logits
sample_image |> n1
perturbed_sample_image |> n1
# checking the perturbation
perturbed_sample_image .- sample_image
# viewing the perturbed image 
colorview(Gray, perturbed_sample_image[1, :, :, 1])
```

I am surprised that this isn't behaving the way we expected, but I haven't executed the notebooks in a few years so it's possible I broke something. Having said that, we do [test this function](https://github.com/vtjeng/MIPVerify.jl/blob/a1d5d06510dcf9682978ebff67d2ba78bcb8518f/test/integration/sequential/trained_weights/MNIST.WK17a_linf0.1_authors.jl#L18), so I am not sure what exactly is going on here.

**Comment by vtjeng** [comment]
One possibility to check first: do the logits for the actual label and one of the other labels line up? That is possible due to the logic that is being fixed in https://github.com/vtjeng/MIPVerify.jl/pull/159.

**Comment by ace-shifu** [comment]
I've written a straightforward code to assess the practical impact of adversarial examples in real attack scenarios. Upon analysis, I observed that for a few `target_index`, Gurobi successfully solves the true counterexamples. However, in the majority of cases, it produces incorrect results. These errors can be categorized into two types:

1. Some `perturbed_sample_image` are not truly adversarial samples; furthermore, their target_predicted_index as same as the unperturbed one.
2. The prediction result of the adversarial sample is neither the classification label of the original sample nor the adversarial label we expect.

````julia
Threads.@threads for target_index in 1:10
    d = MIPVerify.find_adversarial_example(
        n1, 
        sample_image, 
        target_index, 
        Gurobi.Optimizer,
        Dict("OutputFlag" => 0),
        pp = MIPVerify.LInfNormBoundedPerturbationFamily(0.15),
        norm_order = Inf,
        tightening_algorithm = lp,
    )
    print_summary(d)
    perturbed_sample_image = value.(d[:PerturbedInput])
    target_predicted_output = perturbed_sample_image |> n1
    target_predicted_index = target_predicted_output |> MIPVerify.get_max_index
    
    println("\n")
    if target_index == target_predicted_index
        println("\033[32m Info: target: $target_index attack success! \033[0m")
    else
        println("\033[31m Warning target: $target_index attack fails! predicted_target is: $target_predicted_index \033[0m")
    end
    println("\n")
end

![Screenshot from 2024-03-13 20-53-47](https://github.com/vtjeng/MIPVerify.jl/assets/62993420/64b8d1b4-04ba-4256-88b0-5be57655679c)

**Comment by ace-shifu** [comment]
I've written a straightforward code to assess the practical impact of adversarial examples in real attack scenarios. Upon analysis, I observed that for a few `target_index`, Gurobi successfully solves the true counterexamples. However, in the majority of cases, it produces incorrect results. These errors can be categorized into two types:

1. Some `perturbed_sample_image` are not truly adversarial samples; furthermore, their target_predicted_index as same as the unperturbed one.
2. The prediction result of the adversarial sample is neither the classification label of the original sample nor the adversarial label we expect.

````julia
Threads.@threads for target_index in 1:10
    d = MIPVerify.find_adversarial_example(
        n1, 
        sample_image, 
        target_index, 
        Gurobi.Optimizer,
        Dict("OutputFlag" => 0),
        pp = MIPVerify.LInfNormBoundedPerturbationFamily(0.15),
        norm_order = Inf,
        tightening_algorithm = lp,
    )
    print_summary(d)
    perturbed_sample_image = value.(d[:PerturbedInput])
    target_predicted_output = perturbed_sample_image |> n1
    target_predicted_index = target_predicted_output |> MIPVerify.get_max_index
    
    println("\n")
    if target_index == target_predicted_index
        println("\033[32m Info: target: $target_index attack success! \033[0m")
    else
        println("\033[31m Warning target: $target_index attack fails! predicted_target is: $target_predicted_index \033[0m")
    end
    println("\n")
end

![Screenshot from 2024-03-13 20-53-47](https://github.com/vtjeng/MIPVerify.jl/assets/62993420/74788ee8-70bc-4950-80da-be43ee33ea42)

**Comment by ace-shifu** [comment]
![Screenshot from 2024-03-13 20-53-47](https://github.com/vtjeng/MIPVerify.jl/assets/62993420/9255bf11-9a74-4cc0-8e15-245798314bbf)

**Comment by ace-shifu** [comment]
Thank you for your assistance. I believe I have resolved the issue at hand. It is advisable to set the `margin` greater than 0 when invoking ```set_max_indexes```.

**Comment by vtjeng** [comment]
Thanks! Would you be able to do me a favor and run the following code so I can test my hypothesis that the logits were equal in that case?

```
sample_image |> n1

# in a separate cell
perturbed_sample_image |> n1
```

**Comment by ace-shifu** [comment]
[finding_ad_test.pdf](https://github.com/vtjeng/MIPVerify.jl/files/14598173/finding_ad_test.pdf)

**Comment by vtjeng** [comment]
Thanks! To summarize for future readers, solvers like Gurobi can produce output that satisfies constraints [within a tolerance](https://www.gurobi.com/documentation/current/refman/grb_tolerances_and_the_lim.html). If you need the logits for the target category to be strictly greater than any non-target category, you can set a margin (see [example](https://github.com/vtjeng/MIPVerify.jl/blob/2810656489c30208afa54898eb27898e550ea5a0/src/net_components/core_ops.jl#L537)).
